test/functional/normal/put_spec.lua

```diff
@@ -0,0 +1,1012 @@
+            )
+        )
+
+    end)
+
+    describe('blockwise register', function()
+        -- put before with 'p'
+        local blockwise_put_defs = non_dotdefs(normal_command_defs)
+        local test_base = [[
+        Lxtest_stringbine of words 1
+        Ltest_stringbine of words 2
+         test_stringb]]
+
+        local function expect_block_creator(expect_base, conversion_table)
+            return expect_creator(function(e,c) return convert_blockwise(e,c,nil,true) end,
+                                        expect_base, conversion_table)
+        end
+
+        run_test_variations(
+            create_test_defs(
+                blockwise_put_defs,
+                '"bp',
+                create_p_action,
+                test_base,
+                expect_block_creator
+            )
+        )
+    end)
+
+    it('adds correct indentation when put with [p and ]p', function()
+        feed('G>>"a]pix<esc>')
+        -- luacheck: ignore
+        expect([[
+        Line of words 1
+        	Line of words 2
+        	xtest_stringa]])
+        feed('uu"a[pix<esc>')
+        -- luacheck: ignore
+        expect([[
+        Line of words 1
+        	xtest_stringa
+        	Line of words 2]])
+    end)
+
+    describe('linewise paste with autoindent', function()
+        -- luacheck: ignore
+        run_linewise_tests([[
+              Line of words 1
+              	Line of words 2
+              xtest_string"]],
+                'put'
+            ,
+            function()
+                funcs.setline('$', '	Line of words 2')
+                -- Set curswant to '8' to be at the end of the tab character
+                -- This is where the cursor is put back after the 'u' command.
+                funcs.setpos('.', {0, 2, 1, 0, 8})
+                execute('set autoindent')
+            end
+        )
+    end)
+
+    describe('put inside tabs with virtualedit', function()
+        local test_string = [[
+        Line of words 1
+           test_stringx"     Line of words 2]]
+        run_normal_mode_tests(test_string, 'p', function()
+            funcs.setline('$', '	Line of words 2')
+            execute('set virtualedit=all')
+            funcs.setpos('.', {0, 2, 1, 2, 3})
+        end)
+    end)
+
+    describe('put after the line with virtualedit', function()
+        local test_string = [[
+        Line of words 1  test_stringx"
+        	Line of words 2]]
+        run_normal_mode_tests(test_string, 'p', function()
+            funcs.setline('$', '	Line of words 2')
+            execute('set virtualedit=all')
+            funcs.setpos('.', {0, 1, 16, 1, 17})
+        end, true)
+    end)
+
+    describe('Visual put', function()
+        describe('basic put', function()
+            local test_string = [[
+            test_stringx" words 1
+            Line of words 2]]
+            run_normal_mode_tests(test_string, 'v2ep', nil, nil, 'Line of')
+        end)
+        describe('over trailing newline', function()
```

Comment:
I use `describe()` because the `run_test_variations()` function I use (through `run_normal_mode_tests()` in this case) calls `it()` for each of the generated test definitions i.e. `describe('basic put' ...)` groups the set of tests where each of `'p', 'gp', 'P', '".p', '2p', 'gP', ...` is tested in the basic case of putting over a visual selection.

Hence if I use an `it()` at the second level, I get the error `test/functional/normal/put_spec.lua:102: 'it' not supported inside current context block`.

That's also why I don't use `before_each()` very much: most of my tests are generated and run in `run_test_variations()`, so I thought I may as well put the setup function calls in that `it()` call.

It may be a little unclear (my bad), but it's how I make sure every put variation is tested in each situation.